File: app/src/parser.rs
/*******************************************************************************
*   (c) 2021 Zondax GmbH
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use core::mem::MaybeUninit;

use zemu_sys::ViewError;

mod address;
mod asset_id;
mod avm_output;
mod constants;
mod coreth;
mod error;
mod initial_state;
mod inputs;
mod message;
mod network_info;
mod object_list;
mod operations;
mod outputs;
mod pvm_output;
mod subnet_auth;
mod subnet_id;
mod transactions;
mod utils;
mod validator;

pub use address::*;
pub use asset_id::AssetId;
pub use avm_output::AvmOutput;
pub use constants::*;
pub use coreth::{export_tx::ExportTx, import_tx::ImportTx};
pub use error::ParserError;
pub use initial_state::{FxId, InitialState};
pub use inputs::{Input, SECPTransferInput, TransferableInput};
pub use message::AvaxMessage;
pub use network_info::*;
pub use object_list::ObjectList;
pub use outputs::{
    NFTMintOutput, NFTTransferOutput, Output, OutputType, SECPMintOutput, SECPOutputOwners,
    SECPTransferOutput, TransferableOutput,
};
pub use pvm_output::PvmOutput;
pub use subnet_auth::SubnetAuth;
pub use subnet_id::*;
pub use transactions::*;
pub use utils::*;
pub use validator::*;

///This trait defines the interface useful in the UI context
/// so that all the different OperationTypes or other items can handle their own UI
pub trait DisplayableItem {
    /// Returns the number of items to display
    fn num_items(&self) -> usize;

    /// This is invoked when a given page is to be displayed
    ///
    /// `item_n` is the item of the operation to display;
    /// guarantee: 0 <= item_n < self.num_items()
    /// `title` is the title of the item
    /// `message` is the contents of the item
    /// `page` is what page we are supposed to display, this is used to split big messages
    ///
    /// returns the total number of pages on success
    ///
    /// It's a good idea to always put `#[inline(never)]` on top of this
    /// function's implementation
    //#[inline(never)]
    fn render_item(
        &self,
        item_n: u8,
        title: &mut [u8],
        message: &mut [u8],
        page: u8,
    ) -> Result<u8, ViewError>;
}

///This trait defines an useful interface to parse
///objects from bytes.
///this gives different objects in a transaction
///a way to define their own deserilization implementation, allowing higher level objects to generalize the
///parsing of their inner types
pub trait FromBytes<'b>: Sized {
    /// this method is avaliable for testing only, as the preferable
    /// option is to save stack by passing the memory where the object should
    /// store itself
    #[cfg(test)]
    fn from_bytes(input: &'b [u8]) -> Result<(&'b [u8], Self), nom::Err<ParserError>> {
        let mut out = MaybeUninit::uninit();
        let rem = Self::from_bytes_into(input, &mut out)?;
        unsafe { Ok((rem, out.assume_init())) }
    }

    ///Main deserialization method
    ///`input` the input data that contains the serialized form in bytes of this object.
    ///`out` the memory where this object would be stored
    ///
    /// returns the remaining bytes on success
    ///
    /// `Safety` Dealing with uninitialize memory is undefine behavior
    /// even in rust, so implementors should follow the rust documentation
    /// for MaybeUninit and unsafe guidelines.
    ///
    /// It's a good idea to always put `#[inline(never)]` on top of this
    /// function's implementation
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>>;
}

File: app/src/parser.c
/*******************************************************************************
 *   (c) 2018 - 2023 Zondax AG
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 ********************************************************************************/

#include "common/parser.h"

#include <stdio.h>
#include <zxformat.h>
#include <zxmacros.h>
#include <zxtypes.h>

#include "coin.h"
#include "crypto.h"
#include "parser_common.h"
#include "parser_impl.h"
#include "rslib.h"

static zxerr_t parser_allocate();

// This buffer will store parser_state.
// Its size corresponds to ParsedObj (Rust struct)
// Maximum required size: 256 bytes
#define PARSER_BUFFER_SIZE 256
static uint8_t parser_buffer[PARSER_BUFFER_SIZE];

// This function is used to allocate memory for the parser state
zxerr_t parser_allocate(parser_tx_t *parser_state) {
    if (parser_state->len % 4 != 0) {
        parser_state->len += parser_state->len % 4;
    }
    if(parser_state->len > PARSER_BUFFER_SIZE) {
        return zxerr_buffer_too_small;
    }
    if(parser_state->state != NULL) {
        return zxerr_unknown;
    }

    MEMZERO(parser_buffer, PARSER_BUFFER_SIZE);
    parser_state->state = (uint8_t *)&parser_buffer;
    return zxerr_ok;
}

// Initialized parser context and state
parser_error_t parser_init_context(parser_context_t *ctx, const uint8_t *buffer, uint16_t bufferSize) {
    if (bufferSize == 0 || buffer == NULL || ctx == NULL) {
        // Not available, use defaults
        return parser_init_context_empty;
    }

    ctx->offset = 0;
    ctx->buffer = NULL;
    ctx->bufferLen = 0;

    ctx->buffer = buffer;
    ctx->bufferLen = bufferSize;
    return parser_ok;
}

// This function is used for parsing input data.
// internally it takes care of initializing parser state and context.
// some calls to rust are required, specially to get the parsed objecd size.
parser_error_t parser_parse(parser_context_t *ctx, const uint8_t *data, size_t dataLen) {
    ctx->tx_obj.state = NULL;
    ctx->tx_obj.len = 0;
    uint32_t size = 0;

    // initialized data on rust side, this include to get 
    // the size of the object to be parsed according to the 
    // tx type.
    CHECK_ERROR(_parser_init(ctx, data, dataLen, &size));

    if (size == 0) {
        return parser_context_unexpected_size;
    }
    ctx->tx_obj.len = size;

    if(parser_allocate(&ctx->tx_obj) != zxerr_ok) {
        return parser_init_context_empty ;
    }

    // Populate/Initialized parsed object 
    // by parsing the input data.
    parser_error_t err = _parser_read(ctx);

    return err; 
}

// Similates UI flow, this reduces the risk of stack overflow, 
// bad pointer aritmetic and other issues during UI.
parser_error_t parser_validate(const parser_context_t *ctx) {
    zemu_log("parser_validate\n");
    // Iterate through all items to check that all can be shown and are valid
    uint8_t numItems = 0;
    CHECK_ERROR(parser_getNumItems(ctx, &numItems))

    char tmpKey[40] = {0};
    char tmpVal[40] = {0};

    for (uint8_t idx = 0; idx < numItems; idx++) {
        uint8_t pageCount = 0;
        CHECK_ERROR(parser_getItem(ctx, idx, tmpKey, sizeof(tmpKey), tmpVal, sizeof(tmpVal), 0, &pageCount))
    }
    return parser_ok;
}

parser_error_t parser_getNumItems(const parser_context_t *ctx, uint8_t *num_items) {
    if (_getNumItems(ctx, num_items) != parser_ok) {
        return parser_unexpected_number_items;
    }
    if (*num_items == 0) {
        return parser_unexpected_number_items;
    }
    return parser_ok;
}

static void cleanOutput(char *outKey, uint16_t outKeyLen, char *outVal, uint16_t outValLen) {
    MEMZERO(outKey, outKeyLen);
    MEMZERO(outVal, outValLen);
    snprintf(outKey, outKeyLen, "?");
    snprintf(outVal, outValLen, " ");
}

static parser_error_t checkSanity(uint8_t numItems, uint8_t displayIdx) {
    if (displayIdx >= numItems) {
        return parser_display_idx_out_of_range;
    }
    return parser_ok;
}

parser_error_t parser_getItem(const parser_context_t *ctx, uint8_t displayIdx, char *outKey, uint16_t outKeyLen,
                              char *outVal, uint16_t outValLen, uint8_t pageIdx, uint8_t *pageCount) {
    uint8_t numItems = 0;
    CHECK_ERROR(parser_getNumItems(ctx, &numItems))
    CHECK_APP_CANARY()

    CHECK_ERROR(checkSanity(numItems, displayIdx))
    cleanOutput(outKey, outKeyLen, outVal, outValLen);

    // Calls rust getter
    return _getItem(ctx, displayIdx, outKey, outKeyLen, outVal, outValLen, pageIdx, pageCount);
}

parser_error_t parser_getNumItemsBlindSign(uint8_t *num_items) {
    if (_getNumItemsBlindSign(num_items) != parser_ok) {
        return parser_unexpected_number_items;
    }
    if (*num_items == 0) {
        return parser_unexpected_number_items;
    }
    return parser_ok;
}

parser_error_t parser_getItemBlindSign(uint8_t displayIdx, char *outKey, uint16_t outKeyLen,
                                       char *outVal, uint16_t outValLen, uint8_t pageIdx, uint8_t *pageCount) {
    uint8_t numItems = 0;
    CHECK_ERROR(parser_getNumItemsBlindSign(&numItems))
    CHECK_APP_CANARY()

    CHECK_ERROR(checkSanity(numItems, displayIdx))
    cleanOutput(outKey, outKeyLen, outVal, outValLen);

    // Calls rust getter
    return _getItemBlindSign(displayIdx, outKey, outKeyLen, outVal, outValLen, pageIdx, pageCount);
}
